### PR TITLE
[WIP] [Loader] Abstract filesystem resource locator and legacy insecure locator implementation

### PR DESCRIPTION
**Pull request by robfrawley**
| Q | A
| --- | ---
| Branch? | 1.0
| Bug fix? | no
| New feature? | yes
| BC breaks? | no
| Deprecations? | no
| Tests pass? | yes
| Fixed tickets | #846 
| License | MIT
| Doc PR | n/a

Abstract file locator, specifically `FileSystemLocator` and `FileSystemInsecureLocator`, outside of their respective "loader" classes to allow for multiple implementations of resource resolution. This includes an insecure resource locator version that allows symbolic links to be passed and blindly followed with no regard as to where they might point. This fixes environments that rely on deploy tools using symbolic links.

The question is, do we _want_ to offer this functionality? @antoligy @cedricziel @davidromani

### CODE REVIEW

**Comment by robfrawley** [comment]
Note that the `FileSystemInsecureLocator` is missing a comprehensive test suite; this will be added later if the consensus is to add this functionality.

File: DependencyInjection/Compiler/MetadataReaderCompilerPass.php [hunk 1]
```diff
@@ -55,14 +55,18 @@ protected function isDefaultMetadataReader(ContainerBuilder $container)
...
-        $compiler = $container->getCompiler();
-        $formatter = $compiler->getLoggingFormatter();
         $message = 'Automatically replaced Imagine ExifMetadataReader with DefaultMetadataReader; '.
             'you might experience issues with LiipImagineBundle; reason: PHP extension "exif" is missing; solution: '.
             'for advanced metadata extraction install the PHP extension "exif" or set a custom MetadataReader '.
             'through the "liip_imagine.meta_data.reader.class" parameter';
 
-        $compiler->addLogMessage($formatter->format($this, $message));
+        if (method_exists($container, 'log')) {
+            $container->log($message);
```
Comment (robfrawley) [inline DependencyInjection/Compiler/MetadataReaderCompilerPass.php]: New in Symfony 3.3

**Comment by robfrawley** [comment]
Thoughts as to whether we want to offer a file locator implementation that is _less secure_, but allows _symbolic links_? @antoligy @cedricziel

File: Binary/Loader/FileSystemLoader.php [hunk 2]
```diff
@@ -30,78 +32,37 @@ class FileSystemLoader implements LoaderInterface
...
      * @param string[]                  $dataRoots
+     * @param bool                      $secureLocator
      */
-    public function __construct(
-        MimeTypeGuesserInterface $mimeTypeGuesser,
-        ExtensionGuesserInterface $extensionGuesser,
-        $dataRoots
-    ) {
-        $this->mimeTypeGuesser = $mimeTypeGuesser;
+    public function __construct(MimeTypeGuesserInterface $mimeGuesser, ExtensionGuesserInterface $extensionGuesser, $dataRoots, $secureLocator = true)
```
Comment (antoligy) [inline Binary/Loader/FileSystemLoader.php]: Might be worth keeping this to multiple lines for readability?

**Comment by antoligy** [comment]
My opinion is that it's worth adding the InsecureFileLoader, that way we can document that this practice is not recommended, without breaking existing deployments.

Happy for a test suite to appear in a future PR just so we can verify that this works, but I see it should function the same way as before we improved resolution.

File: Binary/Loader/FileSystemLoader.php (hunk 2 above)
Comment (robfrawley) [inline Binary/Loader/FileSystemLoader.php]: Yeah, fair enough. I hate multi-line constructors (unless it's well past 120 characters). That said, I know many are big on it (and an 80 character line limit). I'll revert. ;-) :+1:

**Comment by antoligy** [comment]
Went ahead and resolved the conflict (loving that you can do this now!)- Janky commit but it should be resolved by a squash!

File: Binary/Loader/FileSystemLoader.php (hunk 2 above)
Comment (antoligy) [inline Binary/Loader/FileSystemLoader.php]: I wouldn't mind them so much if we could typehint on primitives...  Grr...  😛

**Comment by robfrawley** [comment]
@antoligy Thanks for the conflict resolution! Since you're on board with this approach, I'm gonna go ahead and add the tests to this PR tonight before merging. Should we release `1.7.2` once this is merged? We have a [decent amount of additions](https://github.com/liip/LiipImagineBundle/compare/1.7.1...1.0) for a patch release, no?

**Comment by antoligy** [comment]
Yeah that sounds like a good idea and I think so, as it addresses probably the only repeat complaint since 1.7.0 was released!

**Comment by robfrawley** [comment]
Right, sounds good. :+1: I'll get the tests implemented. Of note, I was just granted collaborator access to the [avalanche123/Imagine](https://github.com/avalanche123/Imagine) repository today, so that should help us push items that effects this bundle upstream in the future.

**Comment by davidromani** [comment]
I think that it is a very good idea. `FileSystemInsecureLocator` can help me a lot, I prefer to release ASAP.

Thank you so much guys.

**Comment by robfrawley** [comment]
@antoligy Can you give this an additional once-over. I moved some stuff around, created a deprecation layer for `FileSystemLoader::__construct()` so we can change the method signature in `2.0`, moved to services for `FileSystemLocator` implementations, and changed the configuration to an `enum` value. Comprehensive test implemented, as well.

**Comment by robfrawley** [comment]
@antoligy Try to find some time to check this out so we can get `1.7.2` released. I just want another set of eyes before merging this large change, and a fair amount has changed since you "okayed" it.

**Comment by robfrawley** [comment]
@davidromani Before merging this, I'd like to confirm it *does* resolve your symbolic link issue. Can you patch your vendor release of this bundle with the PR and write back confirming proper functionality?

To apply this PR as a patch, you must perform the following:

```bash
# install the `master` from source
cd vendor/liip; rm -fr imagine-bundle
git clone https://github.com/liip/LiipImagineBundle.git imagine-bundle
cd imagine-bundle

# apply this PR as a patch
curl -L https://github.com/liip/LiipImagineBundle/pull/866.diff | patch -p1

# clear your cached assets
rm -fr web/path/to/cached/images/
```

Then enable the new `filesystem_insecure` locator by adding the following to your configuration:

```yml
liip_imagine :
  loaders:
    default:
      filesystem:
        locator: filesystem_insecure
```

**Comment by davidromani** [comment]
@robfrawley Yes, it works! Thank you so much!!

**Comment by antoligy** [comment]
Nice work all! 🙂

**Comment by robfrawley** [comment]
@davidromani The [1.7.2 release](https://github.com/liip/LiipImagineBundle/releases/tag/1.7.2) is live and you can immediately upgrade via a `composer require liip/imagine-bundle:~1.7.2`.

**Pull request merged**